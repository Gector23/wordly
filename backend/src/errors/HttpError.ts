import { StatusCodes } from "http-status-codes";

import { cleanStack } from "#utils/errors";

export class HttpError extends Error {
  status: number;
  isOperational: boolean;

  constructor(message: string, options: { status?: number; cause?: unknown } = {}) {
    if (options.cause) {
      cleanStack(options.cause);
    }

    super(message, { cause: options.cause });
    this.status = options.status ?? StatusCodes.INTERNAL_SERVER_ERROR;
    this.isOperational = options.cause === undefined;

    cleanStack(this);
  }
}
