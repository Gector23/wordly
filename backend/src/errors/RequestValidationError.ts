import { StatusCodes } from "http-status-codes";
import { type ValidationError } from "joi";

import { HttpError } from "./HttpError";

export class RequestValidationError extends HttpError {
  details: string[];

  constructor(joiErrors: ValidationError[]) {
    super("Request validation failed.", { status: StatusCodes.BAD_REQUEST });
    this.details = joiErrors.flatMap(err => err.details.map(d => d.message));
  }
}
