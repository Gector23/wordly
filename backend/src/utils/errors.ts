import { HttpError } from "#errors/HttpError";

export const cleanStack = (err: unknown) => {
  if (err instanceof Error && err.stack) {
    err.stack = err.stack
      .split("\n")
      .filter(line => !line.includes("node_modules/router") && !line.includes("wrapError"))
      .join("\n");
  }
};

export const wrapError = (message: string, err: unknown) => {
  if (err instanceof HttpError && err.isOperational) {
    return err;
  }

  if (err instanceof Error) {
    return new HttpError(message, { cause: err });
  }

  return new HttpError(message);
};
