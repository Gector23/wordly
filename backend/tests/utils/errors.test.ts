import { describe, expect, it } from "@jest/globals";

import { HttpError } from "#errors/HttpError";
import { cleanStack, wrapError } from "#utils/errors";

describe("Error utils", () => {
  describe("cleanStack", () => {
    it("should clean the stack trace", () => {
      const stackRows = [
        "Error: Test error",
        "    at wrapError (src/utils/errors.ts:1:1)",
        "    at Object.<anonymous> (src/utils/errors.ts:1:1)",
        "    at Object.<anonymous> (node_modules/router/index.js:1:1)",
      ];

      const error = new Error("Test error");
      error.stack = stackRows.join("\n");
      cleanStack(error);
      expect(error.stack).not.toContain("node_modules/router");
      expect(error.stack).not.toContain("wrapError");
      expect(error.stack).toMatchSnapshot();
    });
  });

  describe("wrapError", () => {
    it("should not wrap an operational error", () => {
      const error = new HttpError("Operational error", { status: 400 });
      const wrappedError = wrapError("Wrapped operational error", error);
      expect(wrappedError).toBe(error);
    });

    it("should wrap a non-operational error", () => {
      const error = new Error("Non-operational error");
      const wrappedError = wrapError("Wrapped non-operational error", error);
      expect(wrappedError).toBeInstanceOf(HttpError);
      expect(wrappedError.message).toBe("Wrapped non-operational error");
    });

    it("should wrap a non-error object", () => {
      const error = "Non-error object";
      const wrappedError = wrapError("Wrapped non-error object", error);
      expect(wrappedError).toBeInstanceOf(HttpError);
      expect(wrappedError.message).toBe("Wrapped non-error object");
    });
  });
});
