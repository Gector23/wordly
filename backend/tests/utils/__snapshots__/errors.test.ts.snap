// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error utils cleanStack should clean the stack trace 1`] = `
"Error: Test error
    at Object.<anonymous> (src/utils/errors.ts:1:1)"
`;
